sentry-types/src/protocol/profile.rs

```diff
@@ -1,77 +1,143 @@
+    #[serde(serialize_with = "serialize_id")]
+    /// Transaction ID
+    pub id: Uuid,
+    /// Transaction Name
+    pub name: String,
+    /// Trace ID
+    pub trace_id: TraceId,
+    /// Transaction start timestamp in nanoseconds
+    pub relative_start_ns: u64,
+    /// Transaction end timestamp in nanoseconds
+    pub relative_end_ns: u64,
+    /// ID of the thread in which the transaction started
+    #[serde(default)]
+    pub active_thread_id: u64,
+}
 
-#[derive(Debug, Deserialize, Serialize, Clone, PartialEq)]
-/// Represents a Symbol
+#[derive(Debug, Serialize, Deserialize, Clone, Eq, PartialEq)]
+/// Single frame of a Sample
 pub struct RustFrame {
-    /// Raw instruction address
+    /// Instruction address
     pub instruction_addr: String,
 }
 
-#[derive(Debug, Deserialize, Serialize, Clone, PartialEq)]
-/// Represents a Sample
+#[derive(Debug, Serialize, Deserialize, Clone, Eq, PartialEq)]
+/// Single sample of a profile
 pub struct Sample {
-    /// List of symbols
-    pub frames: Vec<RustFrame>,
-    /// The thread name
-    pub thread_name: String,
-    /// The thread id
+    /// ID of the relative stack
+    pub stack_id: u32,
+    /// Thread ID
     pub thread_id: u64,
-    /// Nanoseconds elapsed between when the profiler started and when this sample was collected
-    pub nanos_relative_to_start: u64,
+    /// Timestamp at which this sample was collected
+    pub relative_timestamp_ns: u64,
 }
 
-#[derive(Debug, Default, Deserialize, Serialize, Clone, PartialEq)]
-/// Represents a collected Profile
-pub struct SampledProfile {
-    /// Collection start time in nanoseconds
-    pub start_time_nanos: u64,
-    /// Collection start time in seconds
-    pub start_time_secs: u64,
-    /// Collection duration in nanoseconds
-    pub duration_nanos: u64,
-    /// List of collected samples
-    pub samples: Vec<Sample>,
+#[derive(Debug, Serialize, Deserialize, Clone, Eq, PartialEq)]
+/// Thread metadata
+pub struct ThreadMetadata {
+    #[serde(default, skip_serializing_if = "Option::is_none")]
+    /// Thread name
+    pub name: Option<String>,
 }
 
-#[derive(Debug, Default, Deserialize, Serialize, Clone, PartialEq)]
-/// Represents a Profile Envelope ItemType
+#[derive(Debug, Serialize, Deserialize, Clone, Eq, PartialEq)]
+/// Collected Profile
 pub struct Profile {
-    /// Duration in nanoseconds of the Profile
-    pub duration_ns: u64,
-    /// List of debug images
-    pub debug_meta: DebugMeta,
-    /// Platform is `rust`
-    pub platform: String,
-    /// A string describing the architecture of the CPU that is currently in use
-    /// <https://doc.rust-lang.org/std/env/consts/constant.ARCH.html>
+    /// list of samples in a profile
+    pub samples: Vec<Sample>,
+    /// List of stacks: each stacks is a vec of indexed frames
+    pub stacks: Vec<Vec<u32>>,
+    /// List of frames
+    pub frames: Vec<RustFrame>,
+    /// Thread metadata
+    pub thread_metadata: HashMap<String, ThreadMetadata>,
```

Comment:
what is the key here exactly? I wonder why it is not indexed using a `u32` similar to how stacks/frames works?

Reply:
That's because json keys can only be strings in an object. Otherwise yes, a `u32` would have made more sense.